### ISSUE TITLE
how could I insert a stream of VNodes

### ISSUE THREAD

**Issue by xialvjun** [theadam/react-flyd#1]
Image I have a stream of list --- list$ --- I mean list$, not item$, each one of the stream is an array. 
Then:

``` js
<ul>
  {list$.map(list=>(
    list.map(item=>(
      <li>{item}</li>
    ))
  ))}
</ul>

or just

{list$.map(list=>(
  <ul>
    list.map(item=>(
      <li>{item}</li>
    ))
  </ul>
))}
```

I've viewed the source code, and know `react-flyd` will wrap a stream in a `span`...

But that must be wrong.... Maybe wrap the bottom way in a span is ok(of course the css must change)...

Then, I look into the source code of `react-flyd-class`. 
In it's render method, it `return React.createElement(tag, finalProps);`...Of course it right since React require a valid component just return one VNode...

So, I just think you can offer a function to wrap the stream in a custom tag...

then we can code like

``` js
<div>
  {wrap(
    list$.map(list=>(
      list.map(item=>(
        <li>{item}</li>
      ))
    )), //required
    'ul', //option
    ulProps,  //option
  )}
</div>

or just

<div>
  {wrap(
    {list$.map(list=>(
      <ul>
        list.map(item=>(
          <li>{item}</li>
        ))
      </ul>
    ))},  // since we can return React.onlyChild(children) in a valid React Component's render method, and ul can be the only Child..
    undefined,
    undefined,
  )}
</div>
```

In fact, what I want is just leave the control of wrapChildren to us user rather the framework......
And because this advice need to change `react-flyd` and `react-flyd-class` the two repo, so I don't know how to make a PR....

**Comment by theadam** [comment]
I would like to keep the common case as simple as possible, so the addition of a wrap function would be fine, as long as the example in the README still worked.  This should be almost trivial to implement, i would think.  Even in user-land.

This is untested, but I think wrap can be:

``` js
const wrap = (children, component = 'span', props = {}) => {
  return createElement(reactive(component), props, children);
}
```

This should wrap your stream in a component, which means that when `react-flyd` sees it as a child, `wrapChildren` will just return it directly.

Also, as a side note, I think `react-flyd` will only wrap the child in a span if it is one of many children.  In this example, there is only one child to the `ul` so it seems to function appropriately for me.

**Comment by xialvjun** [comment]
we can make `react-flyd-class` `createReactiveClass` `render` 

```
const finalProps = pickProps(this.state, (key) => key !== 'mount');
if (tag) {
  return React.createElement(tag, finalProps);
} else {
  const {children, ...props} = finalProps
  return React.cloneElement(children, props)
}
```

Then `reactive()` will make **a self control reactive component**, which consume less than **a parent control component**.

And once `isStream(child) && React.isValidElement(child())` , then this child can be a self control reactive component.

Besides, wrapping children in fixed `span` or others will make css confuse, that's not that good.

So I hope using self control reactive component if it can be, otherwise, use parent control component.

Then, I change `react-flyd` `hyperscript`

``` jsx
function wrapChildren(children) {
  const notValidElement$s = children.filter(child => isStream(child) && !isValidElement(child()));
  if (notValidElement$s.length>0) {
    return combine(() => {
      return children.map(child => {
        if (!isStream(child)) {
          return child;
        } else if (notValidElement$s.indexOf(child) > -1) {
          return child();
        } else {
          return createElement(reactive(), {}, child);
        }
      });
    }, notValidElement$s);
  } else {
    return children.map(child => {
      if (!isStream(child)) {
        return child;
      } else {
        return createElement(reactive(), {}, child);
      }
    })
  }
}

export function h(tag, props, ...children) {
  props = props || {}
  children = wrapChildren(children)
  if (hasStream(props) || props && isStream(props.mount) || isStream(children)) {
    if (isStream(children)) {
      return createElement(reactive(tag), toStreams(props), children);
    } else {
      return createElement(reactive(tag), toStreams(props), ...children);
    }
  }
  return createElement(tag, props, ...children);
}
```

Then

``` jsx
<div>{text$} other fixed text</div>
<ul>{list$.map(list=>list.map(item=>(<li key={item}>{item}</li>)))}</ul>
<div>{list$.map(list=>(<ul>{list.map(item=>(<li key={item}>{item}</li>))}</ul>))}<otherdiv /></div>
```

code like above are all ok...

Hope you merge it.

**Comment by theadam** [comment]
Just for ease of communication, considering there are 2 PRs, I will make some comments here.  

First, please run `npm run lint` on the code, I am getting some issues produced there.

Other than that, one thing that worries me about this implementation is the fact the `reactive()` (without `tag`) can only accept one child.  There should be some indication that this is the case, or it should be able to handle multiple children.  I haven't had a chance to look at the rest yet, but that stuck out to me immediately.

**Comment by xialvjun** [comment]
First, many sorry for long time no reply. Then:
I mean no offend, but lint is some time troublesome, it make my code

```
if (isStream(children)) {
  return createElement(reactive(tag), toStreams(props), children);
} else {
  return createElement(reactive(tag), toStreams(props), ...children);
}
```

to

```
let element;
if (isStream(children)) {
  element = createElement(reactive(tag), toStreams(props), children);
} else {
  element = createElement(reactive(tag), toStreams(props), ...children);
}
return element;
```

It's ridiculous verbose.
Well, I'm just got frustrated in using eslint, some problems like:
"TypeError: Cannot read property 'visitClass' of undefined", "not permited to write" and so on... I don't know what's wrong...Yeah, I rarely use eslint.
And now, I've fixed the code to be compatible for eslint and commit them to my fork... And make a new Pull Request. I click "New Pull Request" -> "compare across forks" -> select my fork as head, and then...There is not a 'THEN'.  I saw just a "View Pull Request" button for the old PR.......
Now, there are two "YEAH"s, I rarely use eslint, and I rarely make PRs......
Let it go, let it go, Can't hold it back anymore...(lll￢ω￢)

**Comment by theadam** [comment]
The code should be 

``` js
if (isStream(children)) {
  return createElement(reactive(tag), toStreams(props), children);
}
return createElement(reactive(tag), toStreams(props), ...children);
```

Which is less verbose.  eslint complains here because the `else` statement is unnecessary.

**Comment by xialvjun** [comment]
Well, you are right. And I just found I needn't to make a new PullRequest, since I have an old PR and in it, there are two commits......
As to your 

>  There should be some indication that this is the case, or it should be able to handle multiple children. 

I think if peopel use `react-flyd-class` directly, then since it's himself calling `reactive()`, he should know he couldn't return multiple children... And if people use `react-flyd`, then everything is all invisiable to them. No matter they give a stream of multiple children or a stream of single child, it OK. If it's multiple children, the parent element will handle it; it it's single child, it will handle itself. The difference is **parent control component** consume more than **self control reactive component**.

**Comment by theadam** [comment]
> I think if peopel use react-flyd-class directly, then since it's himself calling reactive(), he should know he couldn't return multiple children

A user would not know this intuitively.  There should be a descriptive error message if they use it incorrectly (rather than the cryptic error they would get from `React.cloneElement`) and there should be an entry in the README.

**Comment by theadam** [comment]
It looks like a lot of changes have been made not directly related to this issue.  So far I have seen that some things related to `mount` were removed from react-flyd-class, and a `toStreams` call on the props passed to `createElement` in `react-flyd`, both of which I believe are necessary to cover the use cases for which this was initially intended. 

Since there are not tests in this project, I do not have the time to adequately manually test all of the cases this project was initially able to handle, and this issue can be resolved with the use of a very simple (one line or so) function a user can create, I cannot merge these pull requests responsibly.

**Comment by theadam** [comment]
Take a look at #4.  This should solve the issue without a special case in `reactive`.  I have added an example that includes the issue you were having.  That PR should close this issue.

**Comment by theadam** [comment]
Fixed in v0.2.1
